File: Data Structures & Algorithms/coin-change-ii/submission-4.py
class Solution:
    def change(self, amount: int, coins: List[int]) -> int:
        dp = [0] * (amount + 1)
        dp[0] = 1

        for i in range(len(coins) - 1, -1, -1):
            nextDp = [0] * (amount + 1)
            nextDp[0] = 1
            for a in range(1, amount + 1):
                nextDp[a] = dp[a]
                if a - coins[i] >= 0:
                    nextDp[a] += nextDp[a - coins[i]]
            dp = nextDp
        return dp[amount]




        # coins.sort()
        # def dfs(i, curr):
        #     if curr == 0:
        #         return 1
        #     if i >= len(coins):
        #         return 0
            
        #     res = 0
        #     if curr >= coins[i]:
        #         res = dfs(i + 1, curr)
        #         res += dfs(i, curr - coins[i])
        #     return res
        
        # return dfs(0, amount)

File: Data Structures & Algorithms/three-integer-sum/submission-4.cpp
class Solution {
public:
    vector<vector<int>> threeSum(vector<int>& nums) {
        vector<vector<int>> res;
        if (nums.size() == 0) return res;

        sort(nums.begin(), nums.end());
        
        for (int i = 0; i < nums.size(); i++) {
            if (i > 0 && nums[i - 1] == nums[i]) continue;
            int l = i + 1;
            int r = nums.size() - 1;

            while(l < r) {
                if ((nums[i] + nums[l] + nums[r]) == 0) {
                    vector<int> temp(3);
                    temp[0] = nums[i];
                    temp[1] = nums[l];
                    temp[2] = nums[r];
                    res.push_back(temp);
                    l++;
                    r--;
                    while (l < r && nums[r] == nums[r+1]) {
                        r--;
                    }
                }
                else if ((nums[i] + nums[l] + nums[r]) > 0) {
                    r--;
                }
                else {
                    l++;
                }
            }

        }
        
        return res;
    }
};
